src/internal.c

```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
+    {
+        if (bin == NULL || hex == NULL || BinSz < 2)
+            return;
+
+        byte HexTbl[16] = { '0','1','2','3','4','5','6','7',
+                            '8','9','a','b','c','d','e','f' };
+
+        byte UpNibble, LwNibble;
+
+        for (int i = BinSz; i > 0; i--) {
+            UpNibble = (*bin >> 4) & 0x0f;
+            LwNibble =  *bin & 0x0f;
+
+            *hex++ = HexTbl[UpNibble];
+            *hex++ = HexTbl[LwNibble];
+            bin++;
+        }
+        *hex = '\0';
+    }
+    /*
+     * This API is an implementation of SessionSecretCB to be passed
+     * to wolfSSL_set_session_secret_cb()
+     *
+     */
+    static int SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                    int* secretSz, void* ctx)
+    {
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        wolfSSL_CTX_keylog_cb_func callback = ssl->ctx->Keylog_cb;
+        if (callback == NULL)
+            return 0;
+
+        /* need to make sure the given master-secret has a meaningful value */
+        int mssz   = *secretSz;
+        int hasVal = 0;
+        for (int i = 0; i < mssz; i++) {
+            if (*((byte*)secret) != 0) {
+                hasVal = 1;
+                break;
+            }
+        }
+        if (hasVal == 0)
+            return 0; /* return sd success */
+
+        /* build up a hex-decoded keylog string
+           "CLIENT_RANDOM <hex-encoded client random> <hex-encoded master-secret>"
+           note that each keylog string does not have LF.
+        */
+        const char* label = "CLIENT_RANDOM";
+        int   labelSz = sizeof("CLIENT_RANDOM");
+        int   buffSz  = labelSz + (RAN_LEN * 2) + 1 + ((*secretSz) * 2) + 1;
+        byte* log     = XMALLOC(buffSz, ssl->heap, DYNAMIC_TYPE_SECRET);
+
+        if (log == NULL)
+            return MEMORY_E;
+
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz -1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+        ToHexString(ssl->arrays->clientRandom,
+                        log + labelSz, RAN_LEN);/* put hex encoded random */
+        XMEMSET(log + labelSz + RAN_LEN * 2, ' ', 1);        /* add space*/
+        ToHexString((byte*)secret, log + labelSz + RAN_LEN*2 +1, *secretSz);
+
+        /* pass the log to the client callback*/
+        callback(ssl, (char*)log);
+
+        XFREE(log, ssl->heap, DYNAMIC_TYPE_SECRET);
+        return 0;
+    }
+#if defined(WOLFSSL_TLS13)
+    static int SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+        const unsigned char* secret, int secretSz, void* ctx)
+    {
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || secretSz == 0)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        wolfSSL_CTX_keylog_cb_func callback = ssl->ctx->Keylog_cb;
```

Comment:
Same as above, consider changing "callback" variable name.  Possibly "logCb".

Reply:
Renamed it.